### ISSUE TITLE
stack overflow too easy with replace!

### ISSUE THREAD

**Issue by KlausC** [JuliaLang/julia#25384]
```
julia> replace!("")

ERROR: StackOverflowError:
Stacktrace:
 [1] #replace!#246(::Int64, ::Function, ::Function, ::String) at ./set.jl:644
 [2] (::getfield(Base, Symbol("#kw##replace!")))(::NamedTuple{(:count,),Tuple{Int64}}, ::typeof(replace!), ::Function, ::String) at ./<missing>:0
 [3] #replace!#246(::Int64, ::Function, ::Function, ::String) at ./set.jl:644
 ```
The same happens with other non-supported argument types. The guilty is in ./set.jl:644. The method immediately calls itself again. Would prefer to see an `ArgumentError`.

Obviously the method definition at line 644
```  
replace!(prednew::Callable, A; count::Integer=typemax(Int)) =
     replace!(prednew, A, count=clamp(count, typemin(Int), typemax(Int)) % Int)
```
has is meant to call a more specific method (with `count::Integer` replaced by `count:Int`. That does not work, if no such more specific method exists (because of the type of `A`).
I would propose to give the functions with `count::Int` a different name (`_replace!`) and so redirect all calls to `replace!(::Callable,::Any;count::Integer)` over this definition.
alternatively:
define `replace!(prednew::Callable, A; count::Integer=typemax(Int)) = throw(MethodError...` to catch the unsupported types of `A`.
PR is in preparation.

**Comment by rfourquet** [comment]
> I would propose to give the functions with count::Int a different name (_replace!) and so redirect all calls to replace!

That's interesting, it was exactly implemented like this initially, but @nalimilan suggested that it's not great that implementations for supporting new collections have to specialize this `_replace!` instead of `replace!`. On the other hand, I thought that it's nice that implementations have to deal only with `count::Int`, and here we are now :) 

> alternatively: define replace!(prednew::Callable, A; count::Integer=typemax(Int)) = throw(MethodError...

Do you mean to have a catch-all `replace!(prednew::Callable, A; count::Int=typemax(Int)) = throw(MethodError.` ?

**Comment by KlausC** [comment]
yep! See the PR. 

> I thought that it's nice that implementations have to deal only with count::Int,<

that should work now.

**Comment by JeffBezanson** [comment]
Fixed by #25386 #25427
